Add on-demand feature groups documentation (#370)
logicalclocks/feature-store-api: docs/templates/on_demand_feature_group.md

# On-Demand (External) Feature Groups

On-demand (External) Feature Groups are Feature Groups which data is stored on an external storage system (e.g. Data Warehouse, S3, ADLS).
From an API prospective, on-demand feature groups can be used in the same way as cached feature groups. Users can pick features from on-demand feature groups to create training datasets. On-demand feature groups can be also used as data source to create derived features, meaning feature on which additional feature engineering is applied.

On-demand feature groups rely on [Storage Connectors](../../integrations/storage-connectors/) to identify the location and authenticate with the external storage.

> When the on-demand feature group is define on top of an external database capabale of running SQL statements (i.e. when using the JDBC, Redshift or Snowflake connectors) the on-demand feature group needs to be defined as a SQL statement. SQL statements can contain feature engineering transformations, when reading the on-demand feature group, the SQL statement is pushed down to the storage which executes it.

Reviewer: ```suggestion
When the on-demand feature group is defined on top of an external database capabale of running SQL statements (i.e. when using the JDBC, Redshift or Snowflake connectors), the on-demand feature group needs to be defined as a SQL statement. SQL statements can contain feature engineering transformations, when reading the on-demand feature group, the SQL statement is pushed down to the storage for execution.
```

        telco_on_dmd.save()
        ```

=== "Scala"

    !!! example "Connecting from Hopsworks"
        ```scala
        val s3Conn = fs.getS3Connector("telco_s3_bucket")
        val telcoOnDmd = (fs.createOnDemandFeatureGroup()
                    .name("telco_s3")
                    .version(1)
                    .dataFormat(OnDemandDataFormat.PARQUET)
                    .description("On-demand feature group for telecom customer data")
                    .storageConnector(s3Conn)
                    .statisticsEnabled(true)
                    .build())
        telcoOnDmd.save()
        ```

## Use cases

There are two use cases in which a user can benefit from on-demand feature groups:

- **Existing feature engineering pipelines**: in case users have recently migrated to Hopsworks Feature Store and they have existing feature engineering pipelines in production. Users can register the output of the existing pipelines as on-demand feature groups in Hopsworks, and immediately use their features to build training datasets. With on-demand feature groups, users do not have to modify the existing pipelines to write to the Hopsworks Feature Store.

> - **Data Ingestion**: on-demand feature groups can be used as a data source. The benefit of using on-demand feature groups to ingest data from external sources is that the Hopsworks Feature Store keeps track of where the data is located and how to authenticate with the external storage system. In addition to that, the Hopsworks Feature Store tracks also the schema of the underlying data will make sure that, if something changes in the underlying schema, the ingestion pipline fails with a clear error.

Reviewer: ```suggestion
- **Data Ingestion**: on-demand feature groups can be used as a data source. The benefit of using on-demand feature groups to ingest data from external sources is that the Hopsworks Feature Store keeps track of where the data is located and how to authenticate with the external storage system. In addition to that, the Hopsworks Feature Store tracks also the schema of the underlying data and will make sure that, if something changes in the underlying schema, the ingestion pipeline fails with a clear error.
```
